improve `--heap-size-hint` arg handling

mbauman opened the pull request:
Previously, `--heap-size-hint` would silently ignore many flavors of "bad" input, parsing things like "3PB" as 3 bytes. This change makes it significantly less permissive, erroring unless it can parse a number (still relying on the C `sscanf` `%Lf` format specifier there) with an optional unit (case-insensitive, either with or without the trailing `b`). Also test it.

mbauman committed:
improve heap-size-hint arg handling

Previously, --heap-size-hint would silently ignore many flavors of "bad" input, parsing things like "3gigs" as 3 bytes. This change makes it significantly less permissive, erroring unless it can parse a number (still relying on the C `sscanf` `%Lf` format specifier there) with an optional unit (case-insensitive, either with or without the trailing `b`). Also test it.

mbauman committed:
more testsets

mbauman committed:
less undefined behavior, more debugging

mbauman committed:
better debugging

mbauman commented on src/jloptions.c:
```diff
@@ -781,33 +782,38 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
+                        jl_errorf("julia: invalid argument to --heap-size-hint (%s)", optarg);
+                        break;
+                }
+                double sz = value * multiplier;
```
```suggestion
                long double sz = value * multiplier;
```

mbauman committed:
long double consistently

KristofferC commented on src/jloptions.c:
```diff
@@ -809,7 +809,11 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
                         break;
                 }
-                jl_options.heap_size_hint = (uint64_t)(value * multiplier);
+                double sz = value * multiplier;
```
Ah, I see. Thanks.

KristofferC commented:
Is this ready to go?

gbaraldi committed:
Merge branch 'master' into mb/heap_size_hint_cleanup

vtjnash commented on test/cmdlineargs.jl:
```diff
@@ -828,3 +828,26 @@ end
+    @testset "--heap-size-hint=$str" for str in ["asdf","","0","1.2vb","b","GB","1.2gb2","42gigabytes","5gig","2GiB","NaNt"]
+        @test errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+        if !errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            @show str
+            try
+                run(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            catch
+            end
```
this test needs to be fixed before we can merge

vtjnash commented on test/cmdlineargs.jl:
```diff
@@ -828,3 +828,26 @@ end
+    exename = `$(Base.julia_cmd())`
+    @test errors_not_signals(`$exename --heap-size-hint -e "exit(0)"`)
+    @testset "--heap-size-hint=$str" for str in ["asdf","","0","1.2vb","b","GB","1.2gb2","42gigabytes","5gig","2GiB","NaNt"]
+        @test errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+        if !errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            @show str
+            try
+                run(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            catch
+            end
+        end
```
IIUC, this is what you were going for?
```suggestion
        heap_size_success = errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
        if !heap_size_success
            @show str
            run(ignorestatus(`$exename --heap-size-hint=$str -e "exit(0)"`))
        end
        @test heap_size_success
```
But I don't think we need the extra debugging code on CI

vtjnash replied:
bump

oscardssmith commented:
Is this ready to merge?

fingolfin commented:
@mbauman do you still have plans to update this, taking the existing comments into account?

vtjnash commented:
Also possibly related to fixes in https://github.com/JuliaLang/julia/pull/52979 and https://github.com/JuliaLang/julia/pull/50843

mbauman commented on test/cmdlineargs.jl:
```diff
@@ -828,3 +828,26 @@ end
+    @test errors_not_signals(`$exename --heap-size-hint -e "exit(0)"`)
+    @testset "--heap-size-hint=$str" for str in ["asdf","","0","1.2vb","b","GB","1.2gb2","42gigabytes","5gig","2GiB","NaNt"]
+        @test errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+        if !errors_not_signals(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            @show str
+            try
+                run(`$exename --heap-size-hint=$str -e "exit(0)"`)
+            catch
+            end
+        end
```
```suggestion
```

mbauman committed:
remove debugging code

mbauman committed:
match verbiage with #50480

mbauman committed:
Merge remote-tracking branch 'origin/master' into HEAD

mbauman commented:
Oh yeah, this was a bit of a fly-by code dump here.  Let's see if we can quickly get this back into shape.  I've removed the debugging test code (tests pass locally; I don't recall whether I was trying to debug something in CI?).  And I've changed the docs to match with https://github.com/JuliaLang/julia/pull/50480.

This looking better?

fingolfin commented on src/jloptions.c:
```diff
@@ -188,8 +189,8 @@ static const char opts[]  =
-    " --heap-size-hint=<size>    Forces garbage collection if memory usage is higher than that value.\n"
-    "                            The memory hint might be specified in megabytes(500M) or gigabytes(1G)\n\n"
+    " --heap-size-hint=<size>    Forces garbage collection if memory usage is greater than the given number of bytes.\n"
+    "                            The memory hint might be specified in megabytes (e.g., 500M) or gigabytes (e.g., 1G)\n\n"
```
To repeat my previous comment: The same text also occurs in `doc/man/julia.1` and with PR https://github.com/JuliaLang/julia/pull/50480 <s>hopefully soon</s> also in `doc/src/manual/command-line-interface.md` so the verbiage should be adjusted to match in those two files, too.

mbauman replied:
Sorry, I didn't mean to blithely dismiss you — I thought I did that in [7495e55](https://github.com/JuliaLang/julia/pull/48050/commits/7495e55a4b729c57933a85018e7dfbbc0b4a4b56)!  Will address

fingolfin replied:
No worries, I assumed it got lost in the shuffle :-)

mbauman committed:
standardize documentation

fingolfin commented on src/jloptions.c:
```diff
@@ -799,33 +800,38 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
+                    jl_errorf("julia: invalid argument to --heap-size-hint (%s)", optarg);
+                }
+                uint64_t multiplier = 1ull;
+                switch (tolower(unit[0])) {
```
So `ctype` is included just for `tolower`, I think. What was wrong with having two `case` per label?

(Not a deal break to me, just wondering *shrug*

mbauman replied:
Yeah, also on line 806 with `tolower(unit[1]) != 'b'`.  This just seems like a net simplification... and we of course do include that header in many other files.

vtjnash replied:
`tolower` seems unreliable since it is locale-specific. Use `utf8proc_tolower` instead

fingolfin commented on test/cmdlineargs.jl:
```diff
@@ -1067,3 +1067,19 @@ if Sys.islinux() && Sys.ARCH in (:i686, :x86_64) # rr is only available on these
+    g = 1024m
+    t = 1024g
+    @testset "--heap-size-hint=$str" for (str, val) in [("1", 1), ("1e7", 1e7), ("2.5e7", 2.5e7), ("1MB", 1m), ("2.5g", 2.5g), ("1e4kB", 1e4k),
+        ("1e100", typemax(UInt64)), ("1e500g", typemax(UInt64)), ("1e-12t", 1), ("500000000b", 500000000)]
```
Hmmm, clearly I did not look closely enough before: I thought this was about adding support for inputs like `2.5g` which seems reasonable to me

But is a heap size hint of the form `1e-12t` or `1e100` really something we want to / should support?

I guess it would be more work to *not* support them, though. Ah well.

mbauman replied:
This PR doesn't change the numeric syntaxes supported here — it just adds tests for them.

The big change here isn't about the supported numeric values, but rather the byte unit.  Before Julia would silently discard/ignore unrecognized units.  After this PR, Julia errors.  That's the big improvement.

Currently:

```
▶ julia --heap-size-hint=2.5GB -E "Base.JLOptions().heap_size_hint"
0x0000000000000002

▶ julia --heap-size-hint=2.5Gibberish -E "Base.JLOptions().heap_size_hint"
0x0000000000000002
```

After:

```
▶ ./julia --heap-size-hint=2.5GB -E "Base.JLOptions().heap_size_hint"
0x00000000a0000000

▶ ./julia --heap-size-hint=2.5Gibberish -E "Base.JLOptions().heap_size_hint"
ERROR: julia: invalid argument to --heap-size-hint (2.5Gibberish)
```

mbauman committed:
use utf8proc_tolower

vtjnash commented:
Apparently something is wrong with `_imp__utf8proc_tolower`. It seems like we must have linked it since `utf8proc_charwidth` exists, but this other one is missing for some reason.

vtjnash committed:
Merge branch 'master' into mb/heap_size_hint_cleanup

vtjnash commented on src/jloptions.c:
```diff
@@ -801,33 +803,38 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
-                    jl_options.heap_size_hint = (uint64_t)(value * multiplier);
+                char unit[4] = {0};
+                int nparsed = sscanf(optarg, "%Lf%3s", &value, unit);
+                if (nparsed == 0 || strlen(unit) > 2 || (strlen(unit) == 2 && utf8proc_tolower(unit[1]) != 'b')) {
```
I don't know how the Windows build explodes so badly with this, so how about we just do this manually:
```suggestion
                if (nparsed == 0 || strlen(unit) > 2 || (strlen(unit) == 2 && ascii_tolower(unit[1]) != 'b')) {
```

```c
static char ascii_tolower(char c)
{
    if ('A' <= c && c <= 'Z')
        return c - 'A' + 'a';
    return c;
}
```

mbauman committed:
use a static ascii_tolower

vtjnash commented:
Also needs a rebase to fix some conflicting doc changes

mbauman committed:
Merge remote-tracking branch 'origin/master' into mb/heap_size_hint_cleanup

mbauman committed:
Merge branch 'master' into mb/heap_size_hint_cleanup

fingolfin commented on src/jloptions.c:
```diff
@@ -18,6 +19,15 @@ char *shlib_ext = ".dylib";
+static char ascii_tolower(char c)
+{
+    if ('A' <= c && c <= 'Z')
+        return c - 'A' + 'a';
```
So, let me recap: this PR changed things to use `tolower`, which has issues, so it was suggested to use `utf8proc_tolower` instead which then was replaced by this hand-written function...

How about my original suggestion to just stick with what the code did before this PR, and simply have two cases in the `switch` for each letter? That seems by far simplest and is guaranteed to have no UTF-8 related issues or anything else and requires no additional explanation?

vtjnash replied:
Haha, yeah. But this works too, and there are a couple extra cases now, including one that is not in a switch statement, so it is not entirely obvious to me that it is significantly better one way or the other; just a style preference.
